spec/unit/stats/key_generator_spec.rb

```diff
@@ -0,0 +1,143 @@
+        ThreeScale::Backend::Stats::Keys.application_response_code_value_key(service_id, app_id, code, period)
+      end
+    end
+  end
+  let(:expected_keys_responsecode_user) do
+    codes = %w[200 2XX 403 404 4XX 500 503 5XX]
+    periods = %i[hour day week month year eternity]
+    code_app_period = codes.product(applications, periods)
+    code_app_period.flat_map do |code, app_id, gr|
+      periods(gr, from, to).flat_map do |period|
+        ThreeScale::Backend::Stats::Keys.application_response_code_value_key(service_id, app_id, code, period)
+      end
+    end
+  end
+  let(:expected_keys_usage_service) do
+    periods = %i[hour day week month eternity]
+    metric_period = metrics.product(periods)
+    metric_period.flat_map do |metric_id, gr|
+      periods(gr, from, to).flat_map do |period|
+        ThreeScale::Backend::Stats::Keys.service_usage_value_key(service_id, metric_id, period)
+      end
+    end
+  end
+  let(:expected_keys_usage_application) do
+    periods = %i[hour day week month year eternity]
+    metric_app_period = metrics.product(applications, periods)
+    metric_app_period.flat_map do |metric_id, app_id, gr|
+      periods(gr, from, to).flat_map do |period|
+        ThreeScale::Backend::Stats::Keys.application_usage_value_key(service_id, app_id, metric_id, period)
+      end
+    end
+  end
+  let(:expected_keys_usage_user) do
+    periods = %i[hour day week month year eternity]
+    metric_user_period = metrics.product(users, periods)
+    metric_user_period.flat_map do |metric_id, user_id, gr|
+      periods(gr, from, to).flat_map do |period|
+        ThreeScale::Backend::Stats::Keys.user_usage_value_key(service_id, user_id, metric_id, period)
+      end
+    end
+  end
+
+  subject { described_class.new(job_params).keys }
+
+  context 'responsecode_service keys' do
+    it 'include expected keys' do
+      is_expected.to include(*expected_keys_responsecode_service)
+    end
+  end
+
+  context 'responsecode_application keys' do
+    let(:applications) { %w[1] }
+
+    it 'include expected keys' do
+      is_expected.to include(*expected_keys_responsecode_application)
+    end
+  end
+
+  context 'responsecode_user keys' do
+    let(:users) { %w[1] }
+
+    it 'include expected keys' do
+      is_expected.to include(*expected_keys_responsecode_user)
+    end
+  end
+
+  context 'usage service keys' do
+    let(:metrics) { %w[1] }
+
+    it 'include expected keys' do
+      is_expected.to include(*expected_keys_usage_service)
+    end
+  end
+
+  context 'usage application keys' do
+    let(:metrics) { %w[1] }
+    let(:applications) { %w[10] }
+
+    it 'include expected keys' do
+      is_expected.to include(*expected_keys_usage_application)
+    end
+  end
+
+  context 'usage user keys' do
+    let(:metrics) { %w[1] }
+    let(:users) { %w[10] }
+
+    it 'include expected keys' do
+      is_expected.to include(*expected_keys_usage_user)
+    end
+  end
+  context 'usage user keys' do
```

Comment:
Would the `contains_exactly` matcher be more appropriate? https://relishapp.com/rspec/rspec-expectations/v/3-8/docs/built-in-matchers/contain-exactly-matcher

Reply:
```

Reply:
Ah right, of course :+1: